app/src/main/java/org/theotech/ceaselessandroid/activity/MainActivity.java

```diff
@@ -121,4 +152,33 @@ public void notify(FragmentState fragmentState) {
         // update current fragment
         currentFragment = fragmentState;
     }
+
+    private void populateContacts() {
+        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_CONTACTS)
+            != PackageManager.PERMISSION_GRANTED) {
+            // Should we show an explanation?
+            if (ActivityCompat.shouldShowRequestPermissionRationale(
+                        this, Manifest.permission.READ_CONTACTS)) {
+                Log.d(TAG, "Should show rationale");
+                // TODO: Display a fragment or something here.
+                // From the developers handbook:
+                // Show an explanation to the user *asynchronously* -- don't block
+                // this thread waiting for the user's response! After the user
+                // sees the explanation, try again to request the permission.
+                final Handler handler = new Handler();
+                handler.postDelayed(new Runnable() {
+                    @Override
+                    public void run() {
+                        populateContacts();
```

Comment:
Adding for posterity:
Looks like it does not ask for permission again and goes into a loop because ActivityCompat.shouldShowRequestPermissionRationale() keeps returning true. And like you said, it does not crash the app, just keeps spawning background threads.

It works fine on L, so I'm just going to open an issue and merge this. Thanks for the contribution!